Fix #3780: [RTL] Image region selection.

veena14cs opened the pull request:
## Explanation

Fix #3780. This PR fixes Image region selection issues in RTL layouts when tapping around the image.

## Screenshot LTR and RTL
Mobile:
<img width="256" alt="Screenshot_1632187231" src="https://user-images.githubusercontent.com/16301028/134005141-efdc8002-eb8f-47dc-bfce-727d7180036b.png">........<img width="256" alt="Screenshot_1632187141" src="https://user-images.githubusercontent.com/16301028/134005234-2447ba43-35c1-452d-9cdd-6aeda8c78215.png">

<img width="144" alt="Screenshot_1632187215" src="https://user-images.githubusercontent.com/16301028/134005284-8668e8e4-c710-46fc-ba37-c29defdb0fad.png">.......<img width="144" alt="Screenshot_1632187192" src="https://user-images.githubusercontent.com/16301028/134005335-1a156c4a-7cc1-4f79-97a9-226832a70fec.png">

## Screenshot for Tests that pass in Expresso after RTL code.

<img width="384" alt="Screenshot (11)" src="https://user-images.githubusercontent.com/16301028/134821611-0ea39242-1304-4230-94b1-abdeb5229ae5.png">

## Screenshot for Tests that fail in Expresso before RTL code.

<img width="384" alt="Screenshot (12)" src="https://user-images.githubusercontent.com/16301028/134821590-a0cd0476-13fb-43f8-aecf-599cfdd91509.png">

**Note**: Not tested on Roboelectric added a todo comment in the test.

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-(A11y)-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

veena14cs committed:
Update selection_interaction_item.xml

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Update selection_interaction_item.xml

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Update return_to_topic_button_item.xml

veena14cs committed:
Update return_to_topic_button_item.xml

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Update image_region_selection_interaction_item.xml

veena14cs committed:
Update image_region_selection_interaction_item.xml

veena14cs committed:
Update image_region_selection_interaction_item.xml

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

anandwana001 commented:
<img src="https://user-images.githubusercontent.com/25057618/133966721-74f3702f-f97b-470c-8ea1-6ae41e530325.png" width=230 height=440 />



This works for me. Nice solution. 🎉

anandwana001 reviewed (approved):
LGTM

anandwana001 reviewed (changes requested):
One thing I missed here is, forcing the parent layout to be ltr, when in rtl the conversation format what we needed for rtl is broken.

veena14cs commented:
> One thing I missed here is, forcing the parent layout to be ltr, when in rtl the conversation format what we needed for rtl is broken.

Parent layout is containing Image only and Images are not been changed to rtl. I didn't get what is getting broken here?

anandwana001 commented:
Please correct me if I am wrong here, Does the conversation UI in the LRT and RTL is correct in the above-mentioned screenshots?

RTL example on conversational UI (difference of start margins for question and image in both LTR and RTL) - https://github.com/oppia/oppia-android/issues/3780#issuecomment-920806461

veena14cs commented:
> Please correct me if I am wrong here, Does the conversation UI in the LRT and RTL is correct in the above-mentioned screenshots?
> 
> RTL example on conversational UI (difference of start margins for question and image in both LTR and RTL) - [#3780 (comment)](https://github.com/oppia/oppia-android/issues/3780#issuecomment-920806461)

Margins are correct but if you see clearly there is white background for the image due to which you are seeing margin difference.

veena14cs commented:
> > Please correct me if I am wrong here, Does the conversation UI in the LRT and RTL is correct in the above-mentioned screenshots?
> > RTL example on conversational UI (difference of start margins for question and image in both LTR and RTL) - [#3780 (comment)](https://github.com/oppia/oppia-android/issues/3780#issuecomment-920806461)
> 
> Margins are correct but if you see clearly there is white background for the image due to which you are seeing margin difference.

Yes there is difference between margins. There was padding so it was showing white background. And I think layout direction doesn't work on `ImageView` to fix this I may need to create one more layout that holds imageview

veena14cs commented:
@anandwana001 I have fixed parent layout please check. I haven't removed padding as I am not sure if I have to keep padding or not. I have updated screenshots and code PTAL.

anandwana001 commented:
The screenshots are now look correct but didn't know why we have to wrap this in constraint layout.

BenHenning commented on app/src/main/res/layout/image_region_selection_interaction_item.xml:
```diff
@@ -11,17 +11,10 @@
 
-  <FrameLayout
-    android:id="@+id/interaction_container_frame_layout"
+  <androidx.constraintlayout.widget.ConstraintLayout
```
Why is a ConstraintLayout being used? There's just one child--it seems unnecessary.

veena14cs replied:
As layout direction is affecting margins if we force it to LTR I added a main container to work for RTL and only layout that hold a child will be changed to LTR.

BenHenning replied:
Looks like this isn't needed anymore & was removed.

BenHenning commented on app/src/main/res/layout/image_region_selection_interaction_item.xml:
```diff
@@ -43,20 +36,38 @@
+      android:background="@drawable/drag_drop_white_background"
+      android:descendantFocusability="beforeDescendants"
+      android:focusableInTouchMode="true"
+      android:layoutDirection="ltr"
```
This & the ltr below seems like the main fix. Why not fix this within the custom view, instead? It seems preferable to have custom views properly support RTL rather than disabling it at the parent.

veena14cs replied:
ImageRegionSelectionInteractionView/ImageView is not supporting layout direction. So I fixed it for the parent layout..  I have removed `android:layoutDirection="ltr"` from `ImageRegionSelectionInteractionView`.

BenHenning replied:
Looks like a solution was found at the view level & this is now resolved.

BenHenning reviewed (changes requested):
Thanks for the quick fix @veena14cs.

Besides what @anandwana001 has mentioned above, my main comment is around whether we can fix this from the view's perspective instead of the layout perspective. Is that possible?

veena14cs committed:
Update image_region_selection_interaction_item.xml

veena14cs commented:
> Thanks for the quick fix @veena14cs.
> 
> Besides what @anandwana001 has mentioned above, my main comment is around whether we can fix this from the view's perspective instead of the layout perspective. Is that possible?

I tried fixing it but co-ordinates are not getting laid-out properly in RTL. The image region co-ordinates are from left-to-right and touch co-ordinates that draws frame are from right-to-left and it looks that the co-ordinates are always towards the right end in RTL which is creating problem.

I tried the solution you discussed with @anandwana001 but it didn't work. So I thought this could be the better solution to force the layout to work as LTR, since we aren't changing images from right-to-left for RTL device.

veena14cs commented:
> > Thanks for the quick fix @veena14cs.
> > Besides what @anandwana001 has mentioned above, my main comment is around whether we can fix this from the view's perspective instead of the layout perspective. Is that possible?
> 
> I tried fixing it but co-ordinates are not getting laid-out properly in RTL. The image region co-ordinates are from left-to-right and touch co-ordinates that draws frame are from right-to-left and it looks that the co-ordinates are always towards the right end in RTL which is creating problem.
> 
> I tried the solution you discussed with @anandwana001 but it didn't work. So I thought this could be the better solution to force the layout to work as LTR, since we aren't changing images from right-to-left for RTL device.

This could get fixed from the customView :). I have updated code PTAL.

rt4914 reviewed (changes requested):
LGTM, thanks. 

Can we add a test case for this to make sure if someone removes it the test case will fail.

Also, add one line comment above your solution explaining why we are fixing this to LTR always.

veena14cs commented:
> LGTM, thanks.
> 
> Can we add a test case for this to make sure if someone removes it the test case will fail.
> 
> Also, add one line comment above your solution explaining why we are fixing this to LTR always.

All the tests are ignored in the `ImageRegionSelectionInteractionViewTest`. Is it only for Roboelectric or Expresso too?

veena14cs commented:
> LGTM, thanks.
> 
> Can we add a test case for this to make sure if someone removes it the test case will fail.
> 
> Also, add one line comment above your solution explaining why we are fixing this to LTR always.

Added test for rtl and added comment explaining the solution.

veena14cs committed:
Update image_region_selection_interaction_item.xml

veena14cs committed:
fixed image region selection

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-rtl-image-region-click

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-rtl-image-region-click

veena14cs committed:
Add tests

veena14cs committed:
Update ImageRegionSelectionInteractionViewTest.kt

veena14cs committed:
Update ImageRegionSelectionInteractionViewTest.kt

anandwana001 commented:
@veena14cs 
`Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing`

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/testing/ImageRegionSelectionInteractionViewTest.kt:
```diff
@@ -144,6 +168,45 @@ class ImageRegionSelectionInteractionViewTest {
   fun testImageRegionSelectionInteractionView_clickRegion3_clickRegion2_region2Clicked() {
     launch(ImageRegionSelectionTestActivity::class.java).use {
       it.onActivity {
+        it.window.decorView.layoutDirection = ViewCompat.LAYOUT_DIRECTION_LTR
```
Rather than updating the ignored test, can we have a new test that specifically tests the issue we are trying to solve in this PR. This is we are not modifying the older tests which are already been ignored. 
Or can we confirm that with the new changes in the ignored test are passing the tests or not?

anandwana001 replied:
For me on local the test `android://org.oppia.android.app.testing.ImageRegionSelectionInteractionViewTest.testImageRegionSelectionInteractionView_rtl_clickRegion3_clickRegion2_region2Clicked()` and `android://org.oppia.android.app.testing.ImageRegionSelectionInteractionViewTest.testImageRegionSelectionInteractionView_rtl_clickRegion3_region3Clicked()` are passing else all are ignored.

veena14cs replied:
> @veena14cs
> `Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing`

I have added tests screenshots for RTL specific. I haven't tested on Roboelectric I am unable to test on Arctic I need to check how to run tests on Roboelectric. Can you please check other ignored tests that pass for roboelectric or not as there is todo comment added for all the testcases that points to the issue[ #1611](https://github.com/oppia/oppia-android/issues/1611).

anandwana001 reviewed (changes requested):
Tests are failing on Robolectric and Bazel 

```
org.oppia.android.app.testing.ImageRegionSelectionInteractionViewTest > testImageRegionSelectionInteractionView_rtl_clickRegion3_clickRegion2_region2Clicked FAILED
    expected: NamedRegionClickedEvent(regionLabel=Region 2)
    but was : org.oppia.android.app.utility.DefaultRegionClickedEvent@57b692f9
        at 
```

```
org.oppia.android.app.testing.ImageRegionSelectionInteractionViewTest > testImageRegionSelectionInteractionView_rtl_clickRegion3_region3Clicked FAILED
    expected: NamedRegionClickedEvent(regionLabel=Region 3)
    but was : org.oppia.android.app.utility.DefaultRegionClickedEvent@2e649950
        at org.oppia.android.app.testing.ImageRegionSelectionInteractionViewTest.testImageRegionSelectionInteractionView_rtl_clickRegion3_region3Clicked(ImageRegionSelectionInteractionViewTest.kt:161)
```

veena14cs commented:
@anandwana001 please check if other tests failures have same error as the new tests. If it has same errors I might have to ignore this for now and update the code. Need to test this once [#1611](https://github.com/oppia/oppia-android/issues/1611) is fixed.

veena14cs committed:
Update ImageRegionSelectionInteractionViewTest.kt

veena14cs committed:
Update ImageRegionSelectionInteractionViewTest.kt

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/testing/ImageRegionSelectionInteractionViewTest.kt:
```diff
@@ -280,6 +281,68 @@ class ImageRegionSelectionInteractionViewTest {
 
+  @Test
+  // TODO(#1611): Fix ImageRegionSelectionInteractionViewTest
+  @Ignore
```
Why are we writing ignored tests? Is it because of #1611 ?

BenHenning replied:
Ah this is unfortunate. It seems like Espresso probably is the only choice then.

BenHenning replied:
Actually, given this can we use OppiaTestRule and ``RunOn(TestPlatform.ESPRESSO)`` here instead of ignoring? We don't actually need to ignore tests failing for specific platforms anymore.

veena14cs commented:
Tests doesn't work on Roboelectric as of now.

anandwana001 commented:
@veena14cs in that case we can make them espresso only.

BenHenning commented on app/src/main/java/org/oppia/android/app/utility/ClickableAreasImage.kt:
```diff
@@ -109,6 +110,11 @@ class ClickableAreasImage(
           imageRect.height().roundToInt()
         )
         val newView = View(it.context)
+        // ClickableArea co-ordinates are not laid-out properly in RTL. The image region co-ordinates are
+        // from left-to-right and touch co-ordinates that draws frame are from right-to-left and it looks
+        // that the co-ordinates are always towards the right end in RTL which is creating problem.Thus to
+        // avoid this situation force layout direction to LTR
```
```suggestion
        // ClickableArea coordinates are not laid-out properly in RTL. The image region coordinates are
        // from left-to-right with an upper left origin and touch coordinates from Android start from the right in RTL mode. Thus, to avoid this situation, force layout direction to LTR in all situations.
```

Cleaned up (needs to be re-wrapped at 100 chars per line though).

BenHenning replied:
Not quite done--please also fix the wrapping.

veena14cs replied:
Done.

BenHenning commented on app/src/main/java/org/oppia/android/app/utility/ClickableAreasImage.kt:
```diff
@@ -109,6 +110,11 @@ class ClickableAreasImage(
+        // from left-to-right and touch co-ordinates that draws frame are from right-to-left and it looks
+        // that the co-ordinates are always towards the right end in RTL which is creating problem.Thus to
+        // avoid this situation force layout direction to LTR
+        ViewCompat.setLayoutDirection(it, ViewCompat.LAYOUT_DIRECTION_LTR)
```
This is a really nice & elegant solution.

veena14cs replied:
Thanks Ben!

BenHenning reviewed (commented):
Thanks @veena14cs! Just had one nit comment, otherwise the PR looks like a really nice solution to the problem.

veena14cs committed:
Update app/src/main/java/org/oppia/android/app/utility/ClickableAreasImage.kt

Co-authored-by: Ben Henning <henning.benmax@gmail.com>

veena14cs commented:
> Thanks @veena14cs! Just had one nit comment, otherwise the PR looks like a really nice solution to the problem.

Accepted the nit changes PTAL thanks.

BenHenning commented:
One follow-up to the documentation update comment thread. PTAL @veena14cs

veena14cs commented:
> One follow-up to the documentation update comment thread. PTAL @veena14cs

Done.

veena14cs committed:
Update ClickableAreasImage.kt

veena14cs committed:
Update ClickableAreasImage.kt

rt4914 reviewed (approved):
LGTM.

For tests I will defer to @anandwana001 Thanks.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-rtl-image-region-click

BenHenning reviewed (commented):
Thanks @veena14cs! LGTM though I think we can & should re-enable the tests. See my comment.

veena14cs commented:
> Thanks @veena14cs! LGTM though I think we can & should re-enable the tests. See my comment.

Thanks @BenHenning I have re-enabled tests and updated screenshots in the PR description.
One test was ignored that is to be moved to Roboelectric as per the message.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/testing/ImageRegionSelectionInteractionViewTest.kt:
```diff
@@ -118,7 +121,7 @@ class ImageRegionSelectionInteractionViewTest {
 
```
You had to set the oppia test rule, else the RunOn annotation will not work.

```
  @get:Rule
  val oppiaTestRule = OppiaTestRule()
```

anandwana001 replied:
`HelpFragmentTest` for reference

veena14cs replied:
I tested it without this Rule and I I wonder how it worked. Added it now it shouldn't break anywhere. Thanks for reference.

BenHenning reviewed (approved):
LGTM from code perspective other than the issue @anandwana001 mentioned.

@anandwana001 deferring to you to verify CI is fully passing before merging.

Thanks for quick fix @veena14cs.

veena14cs committed:
Remove ignore and add RunOn Expresso platform

veena14cs committed:
Update ImageRegionSelectionInteractionViewTest.kt

veena14cs commented:
> LGTM from code perspective other than the issue @anandwana001 mentioned.
> 
> @anandwana001 deferring to you to verify CI is fully passing before merging.
> 
> Thanks for quick fix @veena14cs.

Thanks @BenHenning.

anandwana001 commented:
Assign me once finalize. LGTM for merge.

anandwana001 reviewed (approved):
LGTM

veena14cs commented:
> Assign me once finalize. LGTM for merge.

Thanks :)

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-rtl-image-region-click

veena14cs committed:
Add oppiaTest rule

veena14cs committed:
Update ImageRegionSelectionInteractionViewTest.kt
